controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1161,6 +1188,243 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
+
+	hfs = &metal3v1alpha1.HostFirmwareSettings{}
+	if r.Get(context.TODO(), info.request.NamespacedName, hfs) != nil {
+		// Could not get settings, don't return error as settings may not have been available at provisioner
+		info.log.Info("could not get hostFirmwareSettings", "namespacename", info.request.NamespacedName)
+		return nil, nil
+	}
+
+	if hfs.Status.Settings == nil {
+		// If settings cleared after pivot, recreate map and update settings
+		hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+
+		currentSettings, _, err := prov.GetFirmwareSettings(false)
+		if err != nil {
+			info.log.Info("could not get firmware settings from provisioner")
+			return nil, nil
+		}
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return nil, err
+		}
+	}
+
+	schema := &metal3v1alpha1.FirmwareSchema{}
+	if hfs.Status.FirmwareSchema != nil {
+		if r.Get(context.TODO(), client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema) != nil {
+			info.log.Info("failed to get schema for hostFirmwareSettings", "namespacename", hfs.Status.FirmwareSchema.Namespace, "name", hfs.Status.FirmwareSchema.Name)
+			return hfs, nil
+		}
+	} else {
+		schema = nil
+	}
+
+	// Validate the Spec setting using schema, if available
+	if err = ValidateHostFirmwareSettings(hfs, schema); err != nil {
+		return nil, err
+	}
+
+	return hfs, nil
+}
+
+// Get the firmware settings from the provisioner and create hostFirmwareSettings and FirmwareSchema resources
+func (r *BareMetalHostReconciler) storeHostFirmwareSettings(prov provisioner.Provisioner, info *reconcileInfo) (err error) {
+
+	info.log.Info("hostFirmwareSettings")
+
+	// First check if there is a hostFirmwareSettings resource for this host
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(context.TODO(), info.request.NamespacedName, hfs); err == nil {
+		info.log.Info("found existing hostFirmwareSettings resource")
+
+		if hfs.Status.Settings == nil {
+			// If settings was cleared after pivot recreate map, it will be repopulated below
+			hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+		}
+
+		// Only request the settings as schema was retrieved when resource is created
+		currentSettings, _, err := prov.GetFirmwareSettings(false)
+		if err != nil {
+			info.log.Info("could not get firmware settings from provisioner")
+			return err
+		}
+
+		// Update the resource with settings
+		return r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil)
+	}
+	if !k8serrors.IsNotFound(err) {
+		// Error reading the object
+		return errors.Wrap(err, "could not load host firmware settings")
+	}
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		info.log.Info("could not get firmware settings with schema, trying without")
+		// If host cannot return schema attempt to just get the settings
```

Comment:
How can it happen that we can get BIOS settings, but not a schema? In ironic schema is just a part of settings.

Reply:
Now that the client is set to use the API with microversion 1.74 agreed this shouldn't happen, it was an artifact of my testing with an older microversion.